Copy attestation container

DomAyre opened the pull request:
-------------------------------

### _[PR](https://github.com/microsoft/archived-confidential-sidecar-containers/pull/30) moved here as old repo was archived, original PR description:_

Copy attestation container to this repo from CCF ([961c11e](https://github.com/microsoft/confidential-sidecar-containers/pull/30/commits/961c11e77b1b248c3c15038f3c4ae6a11f983f31))

```
This commit is to copy contents of https://github.com/takuro-sato/CCF/tree/335b7101a5616c67fa796ad7fd74e9711b67314b/attestation-container
to this repository with the commit history.

Pre-requisite:
- Install git-filter-repo (https://github.com/newren/git-filter-repo).
  Using this tool seems the easiest way when you want to move multiple directories
  (cosesign1 and sign1util) to a new repo.
Inside hcsshim/:
- git clone <fork of CCF> (It should be newly cloned)
- git filter-repo --path attestation-container
- git checkout -b filter-attestation-container
- git push origin filter-attestation-container
Inside this repository:
- git remote add ccf <fork of CCF>
- git merge --allow-unrelated-histories ccf/filter-attestation-container
```

- Change dir structure so that it marches with existing dir structure
- Update go.mod because of attestation-container's dependencies
- Run go mod tidy
- Update required go version. It's under discussion [here](https://github.com/microsoft/confidential-sidecar-containers/pull/30#discussion_r1184676261)
- Run go mod vendor
- [Rename 'attestation report endorsement' to 'platform certificate' for consistency](https://github.com/microsoft/confidential-sidecar-containers/pull/30/commits/4dc3ee2e2856421893b4ed7b014c41db9b651ede)
-  Refactor pkg/attest and other packages. Remove duplication

## Test

### Manual test inside Confidential ACI instance:

```
# Prepare for E2E test
go run go run ./cmd/attestation-container/ 

# In another terminal
go test ./...
```

### Deploy to Confidential ACI then run test

This should be in the CI in future

With normal Azure account:

```
az login
az account set --subscription <your subscription>
./docker/attestation-container/deploy_and_test.sh  <AZURE_RESOURCE_GROUP> <DOCKER_REGISTRY_NAME> <DEPLOY_ID (anything)>
```

With service principle:

```
./docker/attestation-container/login_service_principal.sh <AZURE_TENANT_ID> <AZURE_APP_ID> <AZURE_SERVICE_PRINCIPAL_PASSWORD> <AZURE_SUBSCRIPTION_ID>
./docker/attestation-container/deploy_and_test.sh  <AZURE_RESOURCE_GROUP> <DOCKER_REGISTRY_NAME> <DEPLOY_ID (anything)>
```

### Test existing code affected by the refactoring

```
go run ./cmd/skr

# in another terminal
export AttestClientRuntimeData="eyJrZXlzIjpbeyJlIjoiQVFBQiIsImtleV9vcHMiOlsiZW5jcnlwdCJdLCJraWQiOiJOdmhmdXEyY0NJT0FCOFhSNFhpOVByME5QXzlDZU16V1FHdFdfSEFMel93Iiwia3R5IjoiUlNBIiwibiI6InY5NjVTUm15cDh6Ykc1ZU5GdURDbW1pU2VhSHB1akcyYkNfa2VMU3V6dkRNTE8xV3lyVUp2ZWFhNWJ6TW9PMHBBNDZwWGttYnFIaXNvelZ6cGlORExDbzZkM3o0VHJHTWVGUGYyQVBJTXUtUlNyek41NnF2SFZ5SXI1Y2FXZkhXay1GTVJEd0FlZnlOWVJIa2RZWWtnbUZLNDRoaFVkdGxDQUtFdjVVUXBGWmp2aDRpSTlqVkJkR1lNeUJhS1FMaGpJNVdJaC1RRzZaYTVzU3VPQ0ZNbm11eXV2TjVEZmxwTEZ6NTk1U3MtRW9CSVktTmlsNmxDdHZjR2dSLUlialVZSEFPczVhamFtVHpnZU84a3gzVkNFOUhjeUtteVVac2l5aUY2SURScDJCcHkzTkhUakl6N3Rta3BUSHg3dEhuUnRsZkUyRlV2MEI2aV9RWWxfWkE1USJ9XX0="
export AttestClientMAAEndpoint=<Your own MAA endpoint>
bash attest_client.sh
bash ./docker/skr/tests/attest_client.sh
```

## TODOs for future PRs

- Run attestation-container's test in CI.
- Create service principal with enough privilege, and put it as GitHub sercret
- Control access to the new CI. e.g. only pre-registered member can run the CI https://docs.github.com/en/organizations/managing-organization-settings/disabling-or-limiting-github-actions-for-your-organization
- Support Linux kernel 6.1. discussion: https://github.com/microsoft/confidential-sidecar-containers/pull/30#discussion_r1183650544
- As I discussed with Ken, we can add retry mechanism to accessing to MAA like fetching platform certs
- Now there is to naming convention for test Test_* and Test*, I guess we can just Test* since it's golang's recommendation. I leave it for now since it makes the diff even more complicated

KenGordon commented on README.md:
```diff
@@ -38,8 +38,8 @@ to provide a CLA and decorate the PR appropriately (e.g., label, comment). Simpl
 
 # Trademarks
-This project may contain trademarks or logos for projects, products, or services. Authorized use of Microsoft trademarks or logos is subject to and must follow Microsoft’s Trademark & Brand Guidelines. Use of Microsoft trademarks or logos in modified versions of this project must not cause confusion or imply Microsoft sponsorship. Any use of third-party trademarks or logos are subject to those th ...
+This project may contain trademarks or logos for projects, products, or services. Authorized use of Microsoft trademarks or logos is subject to and must follow Microsoft’s Trademark & Brand Guidelines. Use of Microsoft trademarks or logos in modified versions of this project must not cause confusion or imply Microsoft sponsorship. Any use of third-party trademarks or logos is subject to those thi ...
```
As above, this seems to be boiler plate. We have these sorts of things following a process of getting CELA approval to open source the code. That is fairly onerous. I'd not change this unless CELA told me it needed changing.

takuro-sato committed:
Copy attestation container (#16)

* Initial attestation container implementation (#4694)

Co-authored-by: Eddy Ashton <ashton.eddy@gmail.com>
Co-authored-by: Julien Maffre <42961061+jumaffre@users.noreply.github.com>

* Fetch attestation endorsement (#4795)

Co-authored-by: Takuro Sato <takurosato@microsoft.com--username>
Co-authored-by: Julien Maffre <42961061+jumaffre@users.noreply.github.com>

* Bump golang.org/x/sys from 0.3.0 to 0.4.0 in /attestation-container (#4847)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Takuro Sato <79583855+takuro-sato@users.noreply.github.com>
Co-authored-by: Julien Maffre <42961061+jumaffre@users.noreply.github.com>

* Bump google.golang.org/grpc from 1.51.0 to 1.52.0 in /attestation-container (#4848)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Amaury Chamayou <amchamay@microsoft.com>

* Increase gRPC timeout for test (#4879)

* Use Unix domain socket (#4858)

Co-authored-by: Takuro Sato <takurosato@microsoft.com--username>

* Bump google.golang.org/grpc from 1.52.0 to 1.52.3 in /attestation-container (#4918)

* Attestation container: retrieve report endorsements from environment (#4948)

* Attestation container: reduce size of container image (#4955)

* Bump golang.org/x/sys from 0.4.0 to 0.5.0 in /attestation-container (#5002)

* Bump google.golang.org/grpc from 1.52.3 to 1.53.0 in /attestation-container (#5003)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Amaury Chamayou <amchamay@microsoft.com>
Co-authored-by: Takuro Sato <79583855+takuro-sato@users.noreply.github.com>
Co-authored-by: Julien Maffre <42961061+jumaffre@users.noreply.github.com>

* Run 'go get -u' and 'go mod tidy' (#5025)

* Make Attestation Container return UVM_REFERENCE_INFO (#4942)

Co-authored-by: Julien Maffre <42961061+jumaffre@users.noreply.github.com>
Co-authored-by: Dominic Ayre <dominicayre@microsoft.com>
Co-authored-by: Mahati Chamarthy <mahati.chamarthy@gmail.com>

* Bump golang.org/x/sys from 0.5.0 to 0.6.0 in /attestation-container (#5081)

* External executor: use attestation container in end-to-end test (#5082)

* Bump google.golang.org/protobuf from 1.28.1 to 1.29.0 in /attestation-container (#5105)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Takuro Sato <79583855+takuro-sato@users.noreply.github.com>

* Update attestation-container document (#5109)

Co-authored-by: Julien Maffre <42961061+jumaffre@users.noreply.github.com>

* Bump google.golang.org/protobuf from 1.29.0 to 1.29.1 in /attestation-container (#5108)

* Bump google.golang.org/protobuf from 1.29.1 to 1.30.0 in /attestation-container (#5127)

* Bump google.golang.org/grpc from 1.53.0 to 1.54.0 in /attestation-container (#5139)

* Bump golang.org/x/sys from 0.6.0 to 0.7.0 in /attestation-container (#5167)

* Attestation container: Support for reading security context from directory (#5221)

* Organize directory structure

* Run only tests that can run without Confidential ACI in CI

* Add tests to CI

* Update Dockerfile so that it can be built with the new directory structure

* Add script to deploy and test attestation-container

* Rename file

* Split script

* Rename 'attestation report endorsement' to 'platform certificate' for consistency

* Update README.md

* TMP

* Temporary hack to test cert fetcher

* Use `fetchWithRetry` for `CertFetcher`

* Use `CertFetcher` for attestation-container

* Revert "Temporary hack to test cert fetcher"

This reverts commit 1fe85941c54f04c0c4c7073cbb646d4203449df9.

* Remove duplication of UVM info related code

* Delete unused code

* Support fake attestation report

* Add test to info_test.go

* Remove unnecessary log

* Add tags to skip E2E test and SNP VM required tests

* Organize files

* Move function

* Fix report data and host data for skr

* Add lisence to .go and .sh files

* Remove unnecessary comment

* Fix Dockerfile

* Improve the test script for attestation-container

* Minor fixes

* Minor improvement

* Use go 1.19

I didn't have to use `-skip` because I can `-tag` instead.

* Go 1.19

* Use go 1.18 which is used by hcsshim

* Use t.Fatalf for tests

* Improve errors

* Use fake report if /dev/sev is not available for MAA attest

* Add comment

* Update retry configuration to avoid being throttled

* Clarify why UDS is used to communicate with sidecar

* Add documentation on the UVM reference info

* Remove stale comment

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Eddy Ashton <ashton.eddy@gmail.com>
Co-authored-by: Julien Maffre <42961061+jumaffre@users.noreply.github.com>
Co-authored-by: Takuro Sato <takurosato@microsoft.com--username>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Amaury Chamayou <amchamay@microsoft.com>
Co-authored-by: Dominic Ayre <dominicayre@microsoft.com>
Co-authored-by: Mahati Chamarthy <mahati.chamarthy@gmail.com>

DomAyre commented:
PR was merged from Takuro's fork into this branch in order to be able to run image pushing CI

Outstanding comments from original PR:

- [ ] @achamayou: Request to explain why UDS is used to communicate with the sidecar
https://github.com/microsoft/archived-confidential-sidecar-containers/pull/30#discussion_r1251759541
Addressed with https://github.com/microsoft/confidential-sidecar-containers/pull/16/commits/7155f52348b52962dbfeeeaccfdaf61996808bbb

- [ ] @achamayou: Request to add UVM reference info documentation
https://github.com/microsoft/archived-confidential-sidecar-containers/pull/30#discussion_r1251760557
Addressed in https://github.com/microsoft/confidential-sidecar-containers/pull/16/commits/d75c3e7f4ed782429d389621b4627efc903b24f2
(I didn't link to the full C-ACI-Scheme document here since that will live in the currently private examples repo, so added a short summary of the format here)

- [ ] @achamayou: Comment is stale
https://github.com/microsoft/archived-confidential-sidecar-containers/pull/30#discussion_r1251764115
Addressed in https://github.com/microsoft/confidential-sidecar-containers/pull/16/commits/4dba1331b6ace870e8e3d52459699dac6729e4b9
(@takuro-sato Is the following line still necessary?)

DomAyre committed:
Add workflow for pushing attestation image

DomAyre committed:
Simplify image tag generation

takuro-sato commented:
> Is the following line still necessary?)

Yes, the compiled binaries are used to run the test.
(If we want, we can build it somewhere outside the Dockerfile though)

DomAyre committed:
Use universal image as base

DomAyre committed:
Merge branch 'main' into attestation-container

DomAyre committed:
Return to minimal image

DomAyre committed:
Push minimal image to registry

DomAyre committed:
.

DomAyre committed:
.

DomAyre committed:
Only copy essential code to container

DomAyre committed:
Drop go mod download/verify

DomAyre committed:
Merge branch 'main' into attestation-container

takuro-sato reviewed (approved):
deploy_and_test.sh passed against the latest commit
